test/test_persistent_cache.py

```diff
@@ -0,0 +1,116 @@
+from absl.testing import absltest, parameterized
+from concurrent.futures import ProcessPoolExecutor
+import functools
+import os
+import sys
+import tempfile
+
+import torch
+import torch_xla.core.xla_model as xm
+import torch_xla.debug.metrics as met
+import torch_xla.distributed.spmd as xs
+import torch_xla.distributed.xla_multiprocessing as xmp
+import torch_xla.runtime as xr
+
+
+# Wrapper to manage a temporary directory for the wrapped test
+def run_with_tmpdir(f):
+
+  @functools.wraps(f)
+  def run(*args, **kwargs):
+    with tempfile.TemporaryDirectory() as tmpdir:
+      kwargs.setdefault('tmpdir', tmpdir)
+      f(*args, **kwargs)
+
+  return run
+
+
+def _test_spawn(fn, args):
+  # Use a new ProcessPoolExecutor for each test to release device locks.
+  with ProcessPoolExecutor() as pool:
+    pool.submit(fn, *args).result()
+
+
+def _assert_correctness_and_metrics(t, xt, metrics):
+  expected = t + t
+  s = xt + xt
+  xm.mark_step()
+  assert torch.allclose(s.cpu(), expected), \
+    f'Incorrect result! expected {expected}, got {s.cpu()}'
+  for counter, value in metrics.items():
+    actual = met.counter_value(counter)
+    assert actual == value, \
+      f'Unexpected value for counter {counter}: expected {value}, got {actual}'
+
+
+def _mp_test(rank, metrics):
+  # In MP, the cache dir must be different for each process to avoid a race
+  # condition where one process loads the compilation result of another, which
+  # would break the metrics assertion.
+  os.environ['XLA_PERSISTENT_CACHE_PATH'] = \
+    os.path.join(os.environ['XLA_PERSISTENT_CACHE_PATH'], str(rank))
+
+  t = torch.randn(16)
+  xt = t.to(xm.xla_device())
+  _assert_correctness_and_metrics(t, xt, metrics)
+
+
+def _single_device_test(metrics):
+  t = torch.randn(16)
```

Comment:
small suggestion: you can move these test cases into your test class by making them `@staticmethod`s

Reply:
I tried that initially, but for the ProcessPoolExecutor it complained that they weren't pickleable 🥲

Reply:
Hmm, are you sure you didn't use `@classmethod`? We have other tests that use `@staticmethod` to create pickleable test cases

Reply:
Just retried the single device test:

```
class PersistentCacheTest(parameterized.TestCase):
  """
  Test suite to verify compilation cache across processes. Tests will run
  multiple Python subprocesses which use the XLA runtime to populate the cache
  and perform assertions on the metrics generated.
  """

  @staticmethod
  def _single_device_test(metrics):
    t = torch.randn(16)
    xt = t.to(xm.xla_device())
    _assert_correctness_and_metrics(t, xt, metrics)
```

It hit this:

```
Traceback (most recent call last):
  File "/home/ptxla/.local/lib/python3.8/site-packages/absl/testing/parameterized.py", line 321, in bound_param_test
    return test_method(self, *testcase_params)
  File "test_persistent_cache.py", line 23, in run
    f(*args, **kwargs)
  File "test_persistent_cache.py", line 103, in test_persistent_cache
    launch_method(test_fn, ({
  File "test_persistent_cache.py", line 31, in _test_spawn
    pool.submit(fn, *args).result()
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 444, in result
    return self.__get_result()
  File "/usr/local/lib/python3.8/concurrent/futures/_base.py", line 389, in __get_result
    raise self._exception
  File "/usr/local/lib/python3.8/multiprocessing/queues.py", line 239, in _feed
    obj = _ForkingPickler.dumps(obj)
  File "/usr/local/lib/python3.8/multiprocessing/reduction.py", line 51, in dumps
    cls(buf, protocol).dump(obj)
TypeError: cannot pickle 'staticmethod' object
```

The _mp_test also hit this when using staticmethod... I wonder what's going wrong lol